### ISSUE TITLE
execute aws_lambda_invocation only once

### ISSUE THREAD

**Issue by DJAlPee** [hashicorp/terraform-provider-aws#4746]
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Description

After creating an aws_lambda_function I want directly to trigger it for a first initialization. Currently I'm doing this using a provisioner and and some shell commands, which are invoking the the lambda with the AWS CLI.

I've seen there is now an aws_lambda_invocation data source, which does exactly the same like my provisioner. In my case, the disadvantage of a data source is, that the lambda is executed every time I do 
a "terraform plan" or "terraform apply", which is not needed and increases the execution time (and AWS costs).

It would be nice to be able to trigger a lambda function only once

### New or Affected Resource(s)
Option 1
Create:
```hcl
resource "aws_lambda_invocation" "invoke"
```
Option 2
Update:
```hcl
data "aws_lambda_invocation" "invoke"
```
### Potential Terraform Configuration

I would prefer the first option, to create a new ressource type. It would look like the data source:
```hcl
resource "aws_lambda_invocation" "example" {
  function_name = "${aws_lambda_function.lambda_function_test.function_name}"

  input = <<JSON
  {
    "key1": "value1",
    "key2": "value2"
  }
  JSON
}
```

As an alternative an optional attribute in the data source can manage the behavior:
```hcl
data "aws_lambda_invocation" "example" {
  function_name = "${aws_lambda_function.lambda_function_test.function_name}"
  trigger_once  = true

  input = <<JSON
  {
    "key1": "value1",
    "key2": "value2"
  }
  JSON
}
```

### References

No references

**Comment by AlPee-DU** [comment]
One further idea on that:
An additional attribute in the resource like e.g. "hash" (stored only in the state) can be you used for forcing a "recreation" of the invocation resource, which will trigger the invocation call each time, a certain depending resource was changed

**Comment by kkapoor1987** [comment]
Is this becoming kinda issue for us. Is there a way to not run data during apply ?

**Comment by dedunumax** [comment]
Can I work on this? I think first option should be the solution.

**Comment by dedunumax** [comment]
@radeksimko Do you think proposed change is the best way to implement this enhancement?

**Comment by dedunumax** [comment]
Temporarily, I created this community provider. [https://registry.terraform.io/providers/dedunumax/awslambda/latest](https://registry.terraform.io/providers/dedunumax/awslambda/latest) But you might need Terraform 0.13 to try that.

**Comment by mcsloe** [comment]
As a workaround, I did the following:

**Comment by Michenux** [comment]
Any news about this issue ?
What's the recommended way until it's fixed ?

**Comment by dedunumax** [comment]
I have made a PR to fix this. But it is not merged yet.

**Comment by rmccarthy-ellevation** [comment]
Any update on if the PR is merged?

**Comment by geofflancaster** [comment]
Bump. Can this get merged?

**Comment by pascal236** [comment]
Please upvote the new PR, thanks.

**Comment by vibou** [comment]
Please merge this as it is a so important BC Break between 0.12 and 0.13 without any workaround besides this new resource.

**Comment by vibou** [comment]
> As a workaround, I did the following:
> 
> ```
> # run lambda during apply only
> data aws_lambda_invocation run {
>   function_name = "my-lambda"
> 
>   input = <<JSON
> {
>   "APPLY_DATE": "${formatdate("DD MMM YYYY hh:mm ZZZ", timestamp())}",
>  ...
> }
> JSON
> }
> ```

So in the lambda you actually check the event is not empty?

**Comment by breathingdust** [comment]
Hi all 👋  Just letting you know that this is issue is featured on this quarters roadmap. If a PR exists to close the issue a maintainer will review and either make changes directly, or work with the original author to get the contribution merged. If you have written a PR to resolve the issue please ensure the "Allow edits from maintainers" box is checked. Thanks for your patience and we are looking forward to getting this merged soon!

**Comment by JackPott** [comment]
When does the quarter finish?

**Comment by vibou** [comment]
any heads up on this one?

**Comment by YakDriver** [comment]
Thank you for your patience. This functionality has been merged and should be part of the next version.
